File: src/services/codeLens.ts
// Code lenses: a reference count over every type and method declaration in a
// file, and an implementation count over interfaces, abstract classes and
// their abstract methods. Reference counts come from ONE pass over the
// workspace (resolving an identifier is memoized per node); implementation
// counts come from the generation-memoized subtype index, so they are
// transitive (B extends A, A implements I counts B under I).

import type { Checker } from "../compiler/checker.ts";
import { forEachChild } from "../compiler/parser.ts";
import type { Program } from "../compiler/program.ts";
import { declarationName, findMethodImplementations, getSubtypeIndex } from "./subtypes.ts";
import {
  type ClassDeclaration,
  type Identifier,
  type MethodDeclaration,
  type Node,
  type SourceFile,
  type Symbol,
  SyntaxKind,
} from "../compiler/types.ts";
import { isSyntheticUri } from "../workspace.ts";

const LENS_DECLARATIONS = new Set<SyntaxKind>([
  SyntaxKind.ClassDeclaration,
  SyntaxKind.InterfaceDeclaration,
  SyntaxKind.EnumDeclaration,
  SyntaxKind.RecordDeclaration,
  SyntaxKind.AnnotationTypeDeclaration,
  SyntaxKind.MethodDeclaration,
]);

export interface CodeLensEntry {
  /** The declaration's name node (the lens anchors to its range). */
  readonly name: Identifier;
  readonly kind: "references" | "implementations";
  /** Reference or implementation sites, excluding the declaration itself. */
  readonly sites: Node[];
}

function isDeclarationName(node: Identifier, symbol: Symbol): boolean {
  const parent = node.parent;
  return !!parent && parent.symbol === symbol && (parent as { name?: Node }).name === node;
}

function hasAbstractModifier(node: { modifiers?: readonly Node[] }): boolean {
  return (node.modifiers ?? []).some(m => m.kind === SyntaxKind.AbstractKeyword);
}

/** The abstract methods of an interface or abstract class declaration. */
export function abstractMethodsOf(declaration: Node): MethodDeclaration[] {
  const members = (declaration as { members?: readonly Node[] }).members ?? [];
  const isInterface = declaration.kind === SyntaxKind.InterfaceDeclaration;
  return members.filter(
    (m): m is MethodDeclaration =>
      m.kind === SyntaxKind.MethodDeclaration &&
      (isInterface
        ? !(m as MethodDeclaration).body // default/static interface methods have one
        : hasAbstractModifier(m as MethodDeclaration)),
  );
}

export function getCodeLenses(
  program: Program,
  checker: Checker,
  sourceFile: SourceFile,
): CodeLensEntry[] {
  // Reference targets: every type/method declaration in this file.
  const refTargets = new Map<Symbol, CodeLensEntry>();
  const entries: CodeLensEntry[] = [];
  const subtypes = getSubtypeIndex(program);

  const collect = (node: Node): void => {
    const name = (node as { name?: Identifier }).name;
    if (name && node.symbol) {
      if (LENS_DECLARATIONS.has(node.kind) && !refTargets.has(node.symbol)) {
        const entry: CodeLensEntry = { name, kind: "references", sites: [] };
        refTargets.set(node.symbol, entry);
        entries.push(entry);
      }
      const isImplTarget =
        node.kind === SyntaxKind.InterfaceDeclaration ||
        (node.kind === SyntaxKind.ClassDeclaration &&
          hasAbstractModifier(node as ClassDeclaration));
      if (isImplTarget) {
        entries.push({
          name,
          kind: "implementations",
          sites: subtypes
            .allSubtypesOf(node.symbol)
            .map(declarationName)
            .filter((n): n is Identifier => n !== undefined),
        });
        for (const method of abstractMethodsOf(node)) {
          entries.push({
            name: method.name,
            kind: "implementations",
            sites: findMethodImplementations(method, program).map(m => m.name),
          });
        }
      }
    }
    forEachChild(node, child => {
      collect(child);
      return undefined;
    });
  };
  collect(sourceFile);
  if (refTargets.size === 0) return entries;

  // One workspace pass for the reference counts: every resolved identifier that
  // names a target counts, except the declaration's own name. Stub files cannot
  // reference user code.
  for (const uri of program.getAllUris()) {
    if (isSyntheticUri(uri)) continue;
    const file = program.getSourceFile(uri);
    if (!file) continue;
    const visit = (node: Node): void => {
      if (node.kind === SyntaxKind.Identifier) {
        const symbol = checker.resolveName(node as Identifier);
        const entry = symbol ? refTargets.get(symbol) : undefined;
        if (entry && !isDeclarationName(node as Identifier, symbol!)) {
          entry.sites.push(node);
        }
      }
      forEachChild(node, child => {
        visit(child);
        return undefined;
      });
    };
    visit(file);
  }
  return entries;
}

File: togo/internal/services/code_lens.go
package services

// Code lenses: a reference count over every type and method declaration in a
// file, and an implementation count over interfaces, abstract classes and their
// abstract methods. Reference counts come from one pass over the workspace;
// implementation counts come from the generation-memoized subtype index.
// Port of src/services/codeLens.ts.

import "github.com/nikeee/cappu/internal/compiler"

func isLensDeclaration(kind compiler.SyntaxKind) bool {
	switch kind {
	case compiler.ClassDeclaration, compiler.InterfaceDeclaration, compiler.EnumDeclaration,
		compiler.RecordDeclaration, compiler.AnnotationTypeDeclaration, compiler.MethodDeclaration:
		return true
	default:
		return false
	}
}

// CodeLensEntry is one lens anchored to a declaration's name.
type CodeLensEntry struct {
	Name  *compiler.Node // the declaration's name node
	Kind  string         // "references" | "implementations"
	Sites []*compiler.Node
}

func lensIsDeclarationName(node *compiler.Node, symbol *compiler.Symbol) bool {
	parent := node.Parent
	return parent != nil && parent.Symbol == symbol && declName(parent) == node
}

func hasAbstractModifier(node *compiler.Node) bool {
	mods := declModifiers(node)
	if mods == nil {
		return false
	}
	for _, m := range mods.Nodes {
		if m.Kind == compiler.AbstractKeyword {
			return true
		}
	}
	return false
}

// abstractMethodsOf returns the abstract methods of an interface or abstract class.
func abstractMethodsOf(declaration *compiler.Node) []*compiler.Node {
	isInterface := declaration.Kind == compiler.InterfaceDeclaration
	var out []*compiler.Node
	for _, m := range declMembers(declaration) {
		if m.Kind != compiler.MethodDeclaration {
			continue
		}
		if isInterface {
			if m.AsMethodDeclaration().Body == nil { // default/static methods have a body
				out = append(out, m)
			}
		} else if hasAbstractModifier(m) {
			out = append(out, m)
		}
	}
	return out
}

// GetCodeLenses returns reference and implementation lenses for a source file.
func GetCodeLenses(program *compiler.Program, checker *compiler.Checker, sourceFile *compiler.Node) []CodeLensEntry {
	refTargets := map[*compiler.Symbol]int{} // symbol -> index into entries
	var entries []CodeLensEntry
	subtypes := GetSubtypeIndex(program)

	var collect compiler.Visitor
	collect = func(node *compiler.Node) bool {
		name := declName(node)
		if name != nil && node.Symbol != nil {
			if isLensDeclaration(node.Kind) {
				if _, ok := refTargets[node.Symbol]; !ok {
					refTargets[node.Symbol] = len(entries)
					entries = append(entries, CodeLensEntry{Name: name, Kind: "references"})
				}
			}
			isImplTarget := node.Kind == compiler.InterfaceDeclaration ||
				(node.Kind == compiler.ClassDeclaration && hasAbstractModifier(node))
			if isImplTarget {
				var sites []*compiler.Node
				for _, sub := range subtypes.AllSubtypesOf(node.Symbol) {
					if n := DeclarationName(sub); n != nil {
						sites = append(sites, n)
					}
				}
				entries = append(entries, CodeLensEntry{Name: name, Kind: "implementations", Sites: sites})
				for _, method := range abstractMethodsOf(node) {
					var impls []*compiler.Node
					for _, m := range FindMethodImplementations(method, program) {
						impls = append(impls, m.AsMethodDeclaration().Name)
					}
					entries = append(entries, CodeLensEntry{Name: method.AsMethodDeclaration().Name, Kind: "implementations", Sites: impls})
				}
			}
		}
		node.ForEachChild(collect)
		return false
	}
	collect(sourceFile)
	if len(refTargets) == 0 {
		return entries
	}

	for _, uri := range program.GetAllUris() {
		if compiler.IsSyntheticURI(uri) {
			continue
		}
		file := program.GetSourceFile(uri)
		if file == nil {
			continue
		}
		var visit compiler.Visitor
		visit = func(node *compiler.Node) bool {
			if node.Kind == compiler.Identifier {
				symbol := checker.ResolveName(node)
				if symbol != nil {
					if idx, ok := refTargets[symbol]; ok && !lensIsDeclarationName(node, symbol) {
						entries[idx].Sites = append(entries[idx].Sites, node)
					}
				}
			}
			node.ForEachChild(visit)
			return false
		}
		visit(file)
	}
	return entries
}
